worker/upload.go

```diff
@@ -0,0 +1,836 @@
+	}
+
+	// refresh contracts
+	u.updatePool(contracts, bh)
+
+	// keep track of used hosts
+	used = make(map[types.PublicKey]types.FileContractID)
+
+	// build contracts map
+	h2c := make(map[types.PublicKey]types.FileContractID)
+	for _, c := range contracts {
+		h2c[c.HostKey] = c.ID
+	}
+
+	// initialize the encryption key
+	o.Key = object.GenerateEncryptionKey()
+	cr := o.Key.Encrypt(r)
+
+	for {
+		proceed, err := func() (bool, error) {
+			// ensure the upload ctx is cancelled
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			// create span
+			ctx, span := tracing.Tracer.Start(ctx, "uploader.slab")
+			defer span.End()
+
+			// initialize upload
+			id := u.newUpload()
+			defer u.finishUpload(id)
+			span.SetAttributes(attribute.Stringer("id", id))
+
+			// create slab
+			slab := object.Slab{
+				Key:       object.GenerateEncryptionKey(),
+				MinShards: uint8(rs.MinShards),
+			}
+
+			// read slab data
+			buf := make([]byte, slab.Length())
+			length, err := io.ReadFull(io.LimitReader(cr, int64(slab.Length())), buf)
+			if err == io.EOF {
+				return false, nil
+			} else if err != nil && err != io.ErrUnexpectedEOF {
+				return false, err
+			}
+			span.AddEvent("shards read")
+
+			// encode and encrypt the shards
+			shards := make([][]byte, rs.TotalShards)
+			slab.Encode(buf, shards)
+			span.AddEvent("shards encoded")
+			slab.Encrypt(shards)
+			span.AddEvent("shards encrypted")
+
+			// upload the shards
+			slab.Shards, err = u.uploadSectors(ctx, id, shards)
+			if err != nil {
+				return false, err
+			}
+
+			// add the slab
+			o.Slabs = append(o.Slabs, object.SlabSlice{
+				Slab:   slab,
+				Offset: 0,
+				Length: uint32(length),
+			})
+
+			// update used hosts
+			for _, ss := range slab.Shards {
+				if _, ok := used[ss.Host]; !ok {
+					used[ss.Host] = h2c[ss.Host]
+				}
+			}
+			return true, nil
+		}()
+		if err != nil {
+			return object.Object{}, nil, err
+		}
+		if !proceed {
+			break
+		}
+	}
+
+	return o, used, nil
+}
+
+func (u *uploader) uploadSectors(ctx context.Context, id uploadID, shards [][]byte) ([]object.Sector, error) {
+	ctx, span := tracing.Tracer.Start(ctx, "uploader.uploadSectors")
+	defer span.End()
+
+	// prepare state
+	state := &uploadState{
+		maxOverdrive: u.w.uploadMaxOverdrive,
+		pending:      make([]int, 0, len(shards)),
+		overdriving:  make(map[int]struct{}, len(shards)),
+		sectors:      make([]object.Sector, len(shards)),
+		rem:          uint64(len(shards)),
+	}
+
+	// prepare launch function
+	launch := func(job *uploadJob, overdrive bool) error {
+		// set the job's span
+		_, jobSpan := tracing.Tracer.Start(ctx, "uploader.uploadJob")
+		job.requestSpan = jobSpan
+		job.requestSpan.SetAttributes(attribute.Bool("overdrive", overdrive))
+
+		// schedule the job
+		if err := u.enqueue(job); err != nil {
+			job.requestSpan.RecordError(err)
+			return err
+		}
+		state.launched(overdrive)
+		return nil
+	}
+
+	// create a timer to trigger overdrive
```

Comment:
I think it might be worthwhile to experiment with a relative overdrive. 
5s or 3s (which I think you changed it to) only works as long as you don't have a ton of load.
Once you upload 10 files each to 3 workers in parallel you will end up in a place where 100% of your workers take longer than 3s and then your overdrive percentage will spike.

e.g. we could only start counting after a certain percentage of shards has been uploaded already. Maybe we can even include the estimates. e.g. if all workers have a 10s estimate there is no point in overdriving after 3 seconds.

Reply:
Well it's capped by `maxOverdrive` right. I thought about this but I figured any sort of relative overdrive was an over-optimisation at this point. The only thing that will happen when a worker is under 100% load, is that it will schedule 36 jobs for a 30 sector upload, which isn't that big of a deal I think because as soon as the upload succeeds its context will cancel and the overdrive jobs don't necessarily all execute.

I don't mind introducing relative overdrive though, I think we should measure it though rather than adding it without baseline and comparison benchmark.